[Skia] Implement inset shadows

GeorgesStavracas opened the pull request:
#### bcffee080ac979d3b38875f008acdd39798ac7a4
<pre>
[Skia] Implement inset shadows
<a href="https://bugs.webkit.org/show_bug.cgi?id=269714">https://bugs.webkit.org/show_bug.cgi?id=269714</a>

Reviewed by Carlos Garcia Campos.

Move the drop shadow filter creation into a helper method, and refactor
all callers of GraphicsContextSkia::createFillPaint() to manually create
and apply a drop shadow. The helper method returns nullptr if there is
no shadow, or if the shadow is effectively hidden, following Cairo&apos;s
optimization.

Add a selector enum for inset and outset shadows, used by the helper
method aforementioned, and handle the case for both inset and outset
shadows. The outset shadow case is just what was there before.

The inset shadow case is a combination of two image filters. One creates
a black shadow with the drop shadow parameters, and uses it as a mask to
apply a color blend. The color blend is what uses the actual drop shadow
color, and blends it with the SrcIn mode, which effectively inverts the
drop shadow based on the pixel opacity.

The only place in code that interprets GraphicsContext.dropShadow() as
an inset shadow is inside GraphicsContext::fillRectWithRoundedHole(),
which is implemented in this commit using SkCanvas::drawDRRect().

* Source/WebCore/platform/graphics/skia/FontCascadeSkia.cpp:
(WebCore::FontCascade::drawGlyphs):
* Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
(WebCore::GraphicsContextSkia::drawNativeImageInternal):
(WebCore::GraphicsContextSkia::fillPath):
(WebCore::GraphicsContextSkia::createDropShadowFilter const):
(WebCore::GraphicsContextSkia::createFillPaint const):
(WebCore::GraphicsContextSkia::fillRect):
(WebCore::GraphicsContextSkia::fillRoundedRectImpl):
(WebCore::GraphicsContextSkia::fillRectWithRoundedHole):
* Source/WebCore/platform/graphics/skia/GraphicsContextSkia.h:

Canonical link: <a href="https://commits.webkit.org/275093@main">https://commits.webkit.org/275093@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cdc6b0febcab80f13661f9ed293e5793ab10c08b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/40887 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/19900 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/43265 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/43445 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/36978 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/43194 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/22904 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/17231 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/33888 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/41461 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/16846 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/35245 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/14511 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/14615 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/36235 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/44737 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/37100 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/36546 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/40291 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/15702 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/12892 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/38659 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/17321 "Built successfully") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/17372 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/5430 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/16965 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
EWS run on current version of this PR (hash https://github.com/WebKit/WebKit/commit/dcb38a09d48590eff74623bb22ef4c4aaf7a2923)<details>

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/40579 "Build is in progress. Recent messages:Running configuration") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/19592 "Build is in progress. Recent messages:Running configuration") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/42957 "Build is in progress. Recent messages:Running configuration") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/43131 "Build is in progress. Recent messages:Running configuration") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/36668 "Build is in progress. Recent messages:Printed configuration") 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/42886 "Build is in progress. Recent messages:") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/22553 "Build is in progress. Recent messages:") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/16923 "Build is in progress. Recent messages:OS: Sonoma (14.1), Xcode: 15.0") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/5/builds/43131 "Build is in progress. Recent messages:Running configuration") | 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/41153 "Build is in progress. Recent messages:") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/49/builds/22553 "Build is in progress. Recent messages:") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-mac](https://ews-build.webkit.org/#/builders/14/builds/42957 "Build is in progress. Recent messages:Running configuration") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-wpe](https://ews-build.webkit.org/#/builders/5/builds/43131 "Build is in progress. Recent messages:Running configuration") | 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/40449 "Build is in progress. Recent messages:") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/49/builds/22553 "Build is in progress. Recent messages:") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/14/builds/42957 "Build is in progress. Recent messages:Running configuration") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/44405 "Build is in progress. Recent messages:Printed configuration") | 
| [⏳ 🛠 🧪 jsc ](https://ews-build.webkit.org/#/builders/JSC-Tests-EWS "Waiting in queue, processing has not started yet") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-ios](https://ews-build.webkit.org/#/builders/49/builds/22553 "Build is in progress. Recent messages:") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/14/builds/42957 "Build is in progress. Recent messages:Running configuration") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/2/builds/44405 "Build is in progress. Recent messages:Printed configuration") | 
| | [⏳ 🛠 tv ](https://ews-build.webkit.org/#/builders/tvOS-17-Build-EWS "Waiting in queue, processing has not started yet") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/51/builds/16923 "Build is in progress. Recent messages:OS: Sonoma (14.1), Xcode: 15.0") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-gtk](https://ews-build.webkit.org/#/builders/2/builds/44405 "Build is in progress. Recent messages:Printed configuration") | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/17013 "Build is in progress. Recent messages:OS: Sonoma (14.0), Xcode: 15.0") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/14/builds/42957 "Build is in progress. Recent messages:Running configuration") | | 
| | [⏳ 🛠 watch ](https://ews-build.webkit.org/#/builders/watchOS-10-Build-EWS "Waiting in queue, processing has not started yet") | | | 
| | [⏳ 🛠 watch-sim ](https://ews-build.webkit.org/#/builders/watchOS-10-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | | | </details>

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/1f2a71cbc07b32d60ccb1b842f23671b17cbd962. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/24751

carlosgcampos commented on Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
```diff
@@ -306,9 +307,8 @@ SkPaint GraphicsContextSkia::createFillPaint(std::optional<Color> fillColor) con
-    // FIXME: Don't add the effect if the shadow is inset
-    if (hasDropShadow()) {
+    // Drop shadow
+    if (hasDropShadow() && shadowStyle != ShadowStyle::None) {
```
I'm no longer sure it's a good idea to handle this here. We know what operations can have a shadow and what kind of shadow to use in every case, so maybe it would be better to add helper functions to create the SkImageFilter and add it to the SkPaint in the operations after calling createFillPaint().

GeorgesStavracas replied:
Moved it to a helper method

carlosgcampos commented on Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
```diff
@@ -318,7 +318,23 @@ SkPaint GraphicsContextSkia::createFillPaint(std::optional<Color> fillColor) con
+            break;
+        }
+
+        paint.setImageFilter(filter);
```
I see in the cairo case we have an isVisible() to avoid creating the shadow if the color is fully transparent, for example. I also wonder what happens if shadow ignores transforms, we don't handle that case here either, no?

GeorgesStavracas replied:
Applied the same optimization to Skia, and added a FIXME for the ignore transform case

carlosgcampos commented on Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
```diff
@@ -623,11 +639,24 @@ void GraphicsContextSkia::fillRoundedRectImpl(const FloatRoundedRect& rect, cons
         return;
-    notImplemented();
+
+    canvas().drawRect(rect, createFillPaint(color, ShadowStyle::None));
+
+    if (hasDropShadow()) {
+        auto skRect = SkRRect::MakeEmpty();
+        const auto& radii = rrect.radii();
+        SkVector skRadii[4] = {
+            { SkFloatToScalar(radii.topLeft().width()), SkFloatToScalar(radii.topLeft().height()) },
+            { SkFloatToScalar(radii.topRight().width()), SkFloatToScalar(radii.topRight().height()) },
+            { SkFloatToScalar(radii.bottomLeft().width()), SkFloatToScalar(radii.bottomLeft().height()) },
+            { SkFloatToScalar(radii.bottomRight().width()), SkFloatToScalar(radii.bottomRight().height()) } };
+        skRect.setRectRadii(rrect.rect(), skRadii);
+        canvas().drawRRect(skRect, createFillPaint(std::nullopt, ShadowStyle::Inset));
```
Skia has API for this, see SkCanvas::drawDRRect()

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/0e44ba36b9240468a253e862ade2880d06f84c7a. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/24751

carlosgcampos commented on Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
```diff
@@ -284,6 +289,42 @@ void GraphicsContextSkia::strokePath(const Path& path)
     canvas().drawPath(*path.platformPath(), createStrokePaint());
 }
 
+sk_sp<SkImageFilter> GraphicsContextSkia::createDropShadowFilter(ShadowStyle shadowStyle) const
```
This should be called createdDropShadowFilterIfNeeded to make it clear that it can return nullptr if not needed.

GeorgesStavracas replied:
Done

carlosgcampos commented on Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
```diff
@@ -284,6 +289,42 @@ void GraphicsContextSkia::strokePath(const Path& path)
+    if (!hasDropShadow())
+        return nullptr;
+
+    const auto shadow = dropShadow();
```
```suggestion
    const auto& shadow = dropShadow();
```

GeorgesStavracas replied:
Done

carlosgcampos commented on Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
```diff
@@ -284,6 +289,42 @@ void GraphicsContextSkia::strokePath(const Path& path)
+    const auto shadow = dropShadow();
+    ASSERT(shadow);
+
+    FloatSize offset = shadow->offset;
```
We don't need to copy the offset, so either use shadow->offset directly or get a const reference.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/cdc6b0febcab80f13661f9ed293e5793ab10c08b. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/24751

GeorgesStavracas committed:
[Skia] Implement inset shadows
https://bugs.webkit.org/show_bug.cgi?id=269714

Reviewed by Carlos Garcia Campos.

Move the drop shadow filter creation into a helper method, and refactor
all callers of GraphicsContextSkia::createFillPaint() to manually create
and apply a drop shadow. The helper method returns nullptr if there is
no shadow, or if the shadow is effectively hidden, following Cairo's
optimization.

Add a selector enum for inset and outset shadows, used by the helper
method aforementioned, and handle the case for both inset and outset
shadows. The outset shadow case is just what was there before.

The inset shadow case is a combination of two image filters. One creates
a black shadow with the drop shadow parameters, and uses it as a mask to
apply a color blend. The color blend is what uses the actual drop shadow
color, and blends it with the SrcIn mode, which effectively inverts the
drop shadow based on the pixel opacity.

The only place in code that interprets GraphicsContext.dropShadow() as
an inset shadow is inside GraphicsContext::fillRectWithRoundedHole(),
which is implemented in this commit using SkCanvas::drawDRRect().

* Source/WebCore/platform/graphics/skia/FontCascadeSkia.cpp:
(WebCore::FontCascade::drawGlyphs):
* Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
(WebCore::GraphicsContextSkia::drawNativeImageInternal):
(WebCore::GraphicsContextSkia::fillPath):
(WebCore::GraphicsContextSkia::createDropShadowFilter const):
(WebCore::GraphicsContextSkia::createFillPaint const):
(WebCore::GraphicsContextSkia::fillRect):
(WebCore::GraphicsContextSkia::fillRoundedRectImpl):
(WebCore::GraphicsContextSkia::fillRectWithRoundedHole):
* Source/WebCore/platform/graphics/skia/GraphicsContextSkia.h:

Canonical link: https://commits.webkit.org/275093@main
